### PR TITLE
Dropdown : Checkbox support

### PR DESCRIPTION
**Pull request by David-Moreira**

### CODE REVIEW

**Comment by David-Moreira** [comment]
@stsrki do you think regular clicks should change multiple selection? They close the dropdown after all, seems more clear to only allow selection on checkboxes since they do not close the dropdown?

**Comment by stsrki** [comment]
> @stsrki do you think regular clicks should change multiple selection? They close the dropdown after all, seems more clear to only allow selection on checkboxes since they do not close the dropdown?

When in multiple selection mode, the click should toggle the checkbox state, no matter where the user clicks on an item.

**Comment by David-Moreira** [comment]
@stsrki However for regular `Dropdown`, it does make sense that a regular click still closes it, right? But clicking the checkbox selects/unselects?

**Comment by stsrki** [comment]
I would really prefer if we could have a real checkbox instead of fake ones, like on Autocomplete. Even if it is hard to do.

> @stsrki However for regular Dropdown, it does make sense that a regular click still closes it, right? But clicking the checkbox selects/unselects?

I'm divided on this one TBH. Both ways make sense, but I tend to go more along that it needs to toggle on the entire width of an item. Like https://flowbite.com/docs/components/dropdowns/#dropdown-with-checkbox for example.

**Comment by David-Moreira** [comment]
> I would really prefer if we could have a real checkbox instead of fake ones, like on Autocomplete. Even if it is hard to do.

I will see what we can do then for having "real " checkboxes then.

> 
> > @stsrki However for regular Dropdown, it does make sense that a regular click still closes it, right? But clicking the checkbox selects/unselects?
> 
> I'm divided on this one TBH. Both ways make sense, but I tend to go more along that it needs to toggle on the entire width of an item. Like https://flowbite.com/docs/components/dropdowns/#dropdown-with-checkbox for example.

Thing is that on that example the dropdown does not close. maybe it's worth adding an `AutoClose` parameter?
But if we do then we have to think about `Multiple` selection behaviour, does it auto close ever? If this new parameter that means it will `AutoClose` by default.

But I guess clicking the whole "row" selects/unselects makes sense then for both `Dropdown` / `DropdownList`. In fact, whatever is implemented in `Dropdown` check wise, should be how `DropdownList` works.

**Comment by stsrki** [comment]
Clicking outside of the menu or pressing an escape key should close it.

**Comment by David-Moreira** [comment]
@stsrki 
It actually worked out fine. I remember having trouble binding checkboxes, properly and re-rendering out of order. Maybe misused it on the past. 
What do you think in considering changing Autocomplete checkbox to use this new feature in **2.0**?

**Comment by stsrki** [comment]
> What do you think in considering changing Autocomplete checkbox to use this new feature in **2.0**?

Absolutely yes. If there aren't any problems.

File: Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs [hunk 1]
```diff
@@ -108,5 +142,20 @@ public bool Disabled
     /// </summary>
     [CascadingParameter] protected Dropdown ParentDropdown { get; set; }
 
+    /// <summary>
+    /// The dropdown renders a checkbox.
+    /// </summary>
+    [Parameter] public bool Checkbox { get; set; }
```
Comment (stsrki) [inline Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs:148]: Rename to ShowCheckbox?

File: Source/Extensions/Blazorise.Components/DropdownList.razor [hunk 2]
```diff
@@ -28,6 +28,9 @@
         var text = TextField?.Invoke( item );
         var value = ValueField != null ? ValueField.Invoke( item ) : default;
 
-        <DropdownItem @key="@item" Clicked="@HandleDropdownItemClicked" Value="@value">@text</DropdownItem>
+        <DropdownItem @key="@item" Clicked="@HandleDropdownItemClicked" Value="@value"
+        Checkbox="@(SelectionMode == DropdownListSelectionMode.Checkbox)"
```
Comment (stsrki) [inline Source/Extensions/Blazorise.Components/DropdownList.razor:32]: Align with other params.

File: Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs [hunk 3]
```diff
@@ -17,10 +18,23 @@ public partial class DropdownItem : BaseComponent
 
     private bool disabled;
 
+    private bool check;
```
Comment (stsrki) [inline Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs:21]: Renamed to `@checked`?

File: Source/Extensions/Blazorise.Components/Enums/AutocompleteFilter.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Blazorise.Components;
+
+/// <summary>
+/// The Autocomplete's filter mode.
+/// </summary>
+public enum AutocompleteFilter
```
[the remark this answers is not in the document]
Comment (David-Moreira) [reply Source/Extensions/Blazorise.Components/Enums/AutocompleteFilter.cs:6]: What? I haven't touched that? have I haha

[the remark this answers is not in the document]
Comment (David-Moreira) [reply Source/Extensions/Blazorise.Components/Enums/AutocompleteFilter.cs:6]: Ahh this is a new file. `AutocompleteFilter` was moved to its own file. `AutocompleteSelectionMode`, remains. They are distinct features.

File: Source/Extensions/Blazorise.Components/Enums/DropdownListSelectionMode.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Blazorise.Components;
+
+/// <summary>
+/// The DropdownList's selection mode.
+/// </summary>
+public enum DropdownListSelectionMode
```
[the remark this answers is not in the document]
Comment (David-Moreira) [reply Source/Extensions/Blazorise.Components/Enums/DropdownListSelectionMode.cs:6]: Don't think so, `DropdownListSelectionMode` is the same convention naming as `AutocompleteSelectionMode`. Both will exist.

**Comment by stsrki** [comment]
Don't forget to update all `DropdownItem` for each of the providers.

**Tailwind**

image

File: Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs (hunk 3 above)
Comment (David-Moreira) [reply Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs:21]: fair enough, I thought we wouldn't like the necessary @ in `@checked`

**Comment by David-Moreira** [comment]
> Don't forget to update all `DropdownItem` for each of the providers.
> 
> **Tailwind**
> 
> image

Nice catch, I forgot to check if there were dedicated files.

File: Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs (hunk 3 above)
Comment (stsrki) [reply Source/Blazorise/Components/Dropdown/DropdownItem.razor.cs:21]: We already have it in several places. So why not.

**Comment by stsrki** [comment]
For tailwind, we might need to adjust the HTML structure based on https://flowbite.com/docs/components/dropdowns/#dropdown-with-checkbox

image

**Comment by stsrki** [comment]
@David-Moreira  Do you think there is anything else you wish to add to this PR?

**Comment by David-Moreira** [comment]
> @David-Moreira Do you think there is anything else you wish to add to this PR?

We'll if you have added the suggested changes for tailwind. No not really.

**Review by stsrki — APPROVED** [review]

**Pull request merged**